src/binary.rs

```diff
@@ -0,0 +1,322 @@
+                    "macos"
+                } else if cfg!(target_os = "linux") {
+                    "linux"
+                } else {
+                    bail!("unsupported OS")
+                }
+            }
+        })
+    }
+
+    /// Direct URL to the release of an application for download.
+    fn url(&self, version: &str) -> Result<String> {
+        Ok(match self {
+            Self::WasmBindgen => format!(
+                "https://github.com/rustwasm/wasm-bindgen/releases/download/{version}/wasm-bindgen-{version}-x86_64-{target}.tar.gz",
+                version = version,
+                target = self.target()?
+            ),
+            Self::WasmOpt => format!(
+                "https://github.com/WebAssembly/binaryen/releases/download/{version}/binaryen-{version}-x86_64-{target}.tar.gz",
+                version = version,
+                target = self.target()?,
+            ),
+        })
+    }
+}
+
+/// Locate the given application and download it if missing.
+#[tracing::instrument(level = "trace")]
+pub async fn get(app: Application, version: Option<&str>) -> Result<PathBuf> {
+    let version = version.unwrap_or_else(|| app.default_version());
+
+    if let Ok(path) = find_system(app, version).await {
+        tracing::info!(app = app.name(), version = version, "using system installed binary");
+        return Ok(path);
+    }
+
+    let cache_dir = cache_dir()?;
+    let app_dir = cache_dir.join(format!("{}-{}", app.name(), version));
+    let bin_path = app_dir.join(app.path());
+
+    if !is_executable(&bin_path)? {
+        let path = download(app, version).await.context("failed downloading release archive")?;
+        let path2 = path.clone();
+
+        async_std::task::spawn_blocking(move || {
+            let mut file = File::open(path2).context("failed opening downloaded file")?;
+            install(app, &mut file, &app_dir)?;
+
+            Ok::<_, anyhow::Error>(())
+        })
+        .await?;
+
+        std::fs::remove_file(path).context("failed deleting temporary archive")?;
+    }
+
+    Ok(bin_path)
+}
+
+/// Try to find a globally system installed version of the application and ensure it is the needed
+/// release version.
+#[tracing::instrument(level = "trace")]
+async fn find_system(app: Application, version: &str) -> Result<PathBuf> {
+    let path = which::which(app.name())?;
+    let output = Command::new(&path).arg("--version").output().await?;
+
+    ensure!(output.status.success(), "running command failed");
+
+    let text = String::from_utf8_lossy(&output.stdout);
+    let text = text.trim();
+
+    let system_version = match app {
+        Application::WasmBindgen => text.splitn(2, ' ').nth(1).context("missing version")?.to_owned(),
+        Application::WasmOpt => text.splitn(2, ' ').nth(1).context("missing version")?.replace(' ', "_"),
+    };
+
+    if system_version == version {
+        Ok(path)
+    } else {
+        bail!("not found")
+    }
+}
```

Comment:
Given how this is used, and that errors would not actually propagate beyond its call point in `fn get`, I would say the following:
- the signature should be changed to return an `Option<PathBuf>`,
- this should be a method of the `Application` type (moved up to its `impl` block towards the top of this file).

Thoughts?

Reply:
> the signature should be changed to return an Option<PathBuf>.

Can do that quickly. Just for reference: I tried to not hide potential errors outcomes where possible, even if they're not used currently. Idea was that it could be used for future error analysis/debugging in the future by adding log calls here and there.
That said, I don't mind to change to `Option` if you think the error value is not of interest. Alternative would be to simply do a debug log if there is an error, but it might as well just bloat the output. Thought this would be helpful in cases of "I've installed application X on my system already, why is it still downloaded?".

> this should be a method of the Application type (moved up to its impl block towards the top of this file).

Quite the same as I answered in the other comment [here](https://github.com/thedodd/trunk/pull/146#discussion_r633192266). Maybe let's discuss the details over there and once we're at a conclusion I'll apply the same decision here.

Reply:
> Alternative would be to simply do a debug log if there is an error

Yea, I think changing the signature to `Option` makes sense, and logging the error also makes sense. I'm not concerned about bloat in this case, as errors should really only come up when there is actually an error checking the version or other such things which are likely valuable to report.

> Maybe let's discuss the details over there

Responded there. Tl;dr, let's leave this as a top-level fn for now.